Userland/Libraries/LibJS/Runtime/FunctionConstructor.cpp

```diff
@@ -37,32 +38,231 @@ FunctionConstructor::~FunctionConstructor()
+    if (kind == FunctionKind::Generator || kind == FunctionKind::AsyncGenerator)
+        parse_options |= FunctionNodeParseOptions::IsGeneratorFunction;
+    if (kind == FunctionKind::Async || kind == FunctionKind::AsyncGenerator)
+        parse_options |= FunctionNodeParseOptions::IsAsyncFunction;
+
+    // 19. Let parameters be ParseText(! StringToCodePoints(P), parameterSym).
+    i32 function_length = 0;
+    auto parameters_parser = Parser { Lexer { parameters_string } };
+    auto parameters = parameters_parser.parse_formal_parameters(function_length, parse_options);
+
+    // 20. If parameters is a List of errors, throw a SyntaxError exception.
+    if (parameters_parser.has_errors()) {
+        auto error = parameters_parser.errors()[0];
+        return vm.throw_completion<SyntaxError>(global_object, error.to_string());
+    }
+
+    // 21. Let body be ParseText(! StringToCodePoints(bodyString), bodySym).
+    bool contains_direct_call_to_eval = false;
+    auto body_parser = Parser { Lexer { body_string } };
+    // Set up some parser state to accept things like return and yield.
+    body_parser.state().in_function_context = true;
+    if (parse_options & FunctionNodeParseOptions::IsGeneratorFunction)
+        body_parser.state().in_generator_function_context = true;
+    (void)body_parser.parse_function_body(parameters, kind, contains_direct_call_to_eval);
+
+    // 22. If body is a List of errors, throw a SyntaxError exception.
+    if (body_parser.has_errors()) {
+        auto error = body_parser.errors()[0];
+        return vm.throw_completion<SyntaxError>(global_object, error.to_string());
+    }
+
+    // 23. NOTE: The parameters and body are parsed separately to ensure that each is valid alone. For example, new Function("/*", "*/ ) {") is not legal.
+    // 24. NOTE: If this step is reached, sourceText must have the syntax of exprSym (although the reverse implication does not hold). The purpose of the next two steps is to enforce any Early Error rules which apply to exprSym directly.
+
+    // 25. Let expr be ParseText(sourceText, exprSym).
+    auto source_parser = Parser { Lexer { source_text } };
+    // This doesn't need any parse_options, it determines those & the function type based on the tokens that were found.
+    auto expr = source_parser.parse_function_node<FunctionExpression>();
+
+    // 26. If expr is a List of errors, throw a SyntaxError exception.
+    if (source_parser.has_errors()) {
+        auto error = source_parser.errors()[0];
         return vm.throw_completion<SyntaxError>(global_object, error.to_string());
     }
 
+    // 27. Let proto be ? GetPrototypeFromConstructor(newTarget, fallbackProto).
+    auto* prototype = TRY(get_prototype_from_constructor(global_object, *new_target, fallback_prototype));
+
+    // 28. Let realmF be the current Realm Record.
+    auto* realm = vm.current_realm();
+
+    // 29. Let scope be realmF.[[GlobalEnv]].
+    Environment* scope = &realm->global_environment();
```

Comment:
Any specific reason for explicitly naming the type?

Reply:
Nope, initially didn't compile with just auto but I think it was a missing include. Updated.